### PR TITLE
Add guard future extension

### PR DESCRIPTION
**Pull request by niazoff**
Mimics the simple `guard` statement in Swift. Useful for verifying a condition of the future before moving forward.

Example usage:
```swift
.guard({ order.userID == user.id }, else: AccessError.unauthorized)
```

### Checklist

- [x] Circle CI is passing (code compiles and passes tests).
- [x] There are no breaking changes to public API.
- [x] New test cases have been added where appropriate.
- [x] All new code has been commented with doc blocks `///`.

### CODE REVIEW

File: Sources/NIOKit/FutureExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// MARK: - Guard
+
+public extension EventLoopFuture {
+	/// Guards that the future's value satisfies the callback's condition or fails with the given error.
+	func `guard`(_ callback: @escaping ((Value) -> Bool), else error: Error) -> EventLoopFuture<Value> {
```
Comment (calebkleveter) [inline Sources/NIOKit/FutureExtensions.swift]: The error that is passed in should be marked `@autoclosure`. This prevents it from bring initialized unless it is actually used.

File: Sources/NIOKit/FutureExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+		let promise = eventLoop.makePromise(of: Value.self)
+		whenSuccess {
+			if callback($0) { promise.succeed($0) } else { promise.fail(error) }
```
Comment (tanner0101) [inline Sources/NIOKit/FutureExtensions.swift]: spacing is off here

File: Sources/NIOKit/FutureExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+		}
+		whenFailure { promise.fail($0) }
```
Comment (tanner0101) [inline Sources/NIOKit/FutureExtensions.swift]: i think using `whenComplete` and switching on the result would be more performant than adding two closures to the future.

File: Tests/NIOKitTests/FutureExtensionsTests.swift [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+	}
+    
+    override func tearDown() {
+        XCTAssertNoThrow(try group.syncShutdownGracefully())
+        group = nil
+	}
+	
+	func testGuard() {
+		let future1 = eventLoop.makeSucceededFuture(1)
```
Comment (tanner0101) [inline Tests/NIOKitTests/FutureExtensionsTests.swift]: spacing off here, and other parts of this file

File: Tests/NIOKitTests/FutureExtensionsTests.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import XCTest
+import NIO
+@testable import NIOKit
+
+final class FutureExtensionsTests: XCTestCase {
```
Comment (tanner0101) [inline Tests/NIOKitTests/FutureExtensionsTests.swift]: this file needs to be added to `XCTTestMain.swift`

File: Sources/NIOKit/FutureExtensions.swift [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// MARK: - Guard
+
+public extension EventLoopFuture {
+    /// Guards that the future's value satisfies the callback's condition or fails with the given error.
+    func `guard`(_ callback: @escaping ((Value) -> Bool), else error: @escaping @autoclosure () -> Error) -> EventLoopFuture<Value> {
+        let promise = eventLoop.makePromise(of: Value.self)
+        whenComplete { result in
+            switch result {
+            case .success(let value):
+                if callback(value) { promise.succeed(value) }
```
Comment (tanner0101) [inline Sources/NIOKit/FutureExtensions.swift]: vapor style prefers having curly braces on new lines.

File: Sources/NIOKit/FutureExtensions.swift [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// MARK: - Guard
+
+public extension EventLoopFuture {
+    /// Guards that the future's value satisfies the callback's condition or fails with the given error.
+    func `guard`(_ callback: @escaping ((Value) -> Bool), else error: @escaping @autoclosure () -> Error) -> EventLoopFuture<Value> {
+        let promise = eventLoop.makePromise(of: Value.self)
+        whenComplete { result in
```
Comment (tanner0101) [inline Sources/NIOKit/FutureExtensions.swift]: `self.` should be used to reference member vars and methods

File: Sources/NIOKit/FutureExtensions.swift [hunk 8]
```diff
@@ -0,0 +1,17 @@
+// MARK: - Guard
```
Comment (tanner0101) [inline Sources/NIOKit/FutureExtensions.swift]: If you put this inside of the extension, it should generate a nice header in our API docs

File: Sources/NIOKit/FutureExtensions.swift [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+public extension EventLoopFuture {
+    /// Guards that the future's value satisfies the callback's condition or fails with the given error.
```
Comment (tanner0101) [inline Sources/NIOKit/FutureExtensions.swift]: example usage in this doc block would be nice for people viewing the API docs

File: Tests/NIOKitTests/FutureExtensionsTests.swift [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+		let future1 = eventLoop.makeSucceededFuture(1)
+		let guardedFuture1 = future1.guard({ $0 == 1 }, else: TestError.notEqualTo1)
+		XCTAssertNoThrow(try guardedFuture1.wait())
+		
+		let future2 = eventLoop.makeSucceededFuture("foo")
+		let guardedFuture2 = future2.guard({ $0 == "bar" }, else: TestError.notEqualToBar)
+		XCTAssertThrowsError(try guardedFuture2.wait())
+	}
+	
+	static var allTests = [("testGuard", testGuard)]
```
Comment (tanner0101) [inline Tests/NIOKitTests/FutureExtensionsTests.swift]: each tuple should be on its own line with commas after (to minimize merge conflicts)

Comment (niazoff) [reply Tests/NIOKitTests/FutureExtensionsTests.swift]: Comma even if only one like `("testGuard", testGuard),`?

Comment (grundoon) [reply Tests/NIOKitTests/FutureExtensionsTests.swift]: Yup... separating each element (including the brackets) to its own line makes for a cleaner merge situation:

```swift
static var allTests = [
    ("testGuard", testGuard),
]
```

File: Tests/NIOKitTests/FutureExtensionsTests.swift [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import XCTest
+import NIO
+@testable import NIOKit
+
+final class FutureExtensionsTests: XCTestCase {
```
Comment (calebkleveter) [inline Tests/NIOKitTests/FutureExtensionsTests.swift]: This should be conformed to `NIOTestCase`. Then lines 6 - 19 can be removed.

Comment (MrLotU) [reply Tests/NIOKitTests/FutureExtensionsTests.swift]: Good catch! :)

**Comment by penny-coin** [comment]
Hey @niazoff, you just merged a pull request, have a coin! 

You now have 1 coins.

**Pull request merged**